cpp/include/cuvs/cluster/agglomerative.hpp

```diff
@@ -119,6 +121,34 @@ void single_linkage(
   cuvs::cluster::agglomerative::Linkage linkage = cuvs::cluster::agglomerative::Linkage::KNN_GRAPH,
   std::optional<int> c                          = std::make_optional<int>(DEFAULT_CONST_C));
 
+/**
+ * Given a neighbors graph, constructs a linkage over it by computing
+ * the minimum spanning tree and dendrogram. Returns mst edges sorted by weight and the linkage.
+ * @tparam value_idx
+ * @tparam value_t
+ * @param[in] handle raft handle for resource reuse
+ * @param[in] X data points (size m * n)
+ * @param[in] metric distance metric to use
+ * @param[in] graph_indptr CSR indices of graph nodes (size m + 1)
+ * @param[in] graph input graph
+ * @param[out] out_mst output MST sorted by edge weights (size m - 1)
+ * @param[out] dendrogram output dendrogram (size [n_rows - 1] * 2)
+ * @param[out] out_distances distances for output
+ * @param[out] out_sizes cluster sizes of output
+ * @param[in] reduction_op reduction operation for computing nearest neighbors while connecting
+ * graph components.  The reduction operation must have `gather` and `scatter` functions defined
+ */
+template <typename red_op>
+void build_linkage(raft::resources const& handle,
+                   raft::device_matrix_view<const float, int, raft::row_major> X,
+                   cuvs::distance::DistanceType metric,
+                   raft::device_vector_view<int, int> graph_indptr,
+                   raft::device_coo_matrix_view<float, int, int, size_t> graph,
+                   raft::device_coo_matrix_view<float, int, int, int> out_mst,
+                   raft::device_matrix_view<int, int> dendrogram,
+                   raft::device_vector_view<float, int> out_distances,
+                   raft::device_vector_view<int, int> out_sizes,
+                   red_op reduction_op);
```

Comment:
We can't compile the reduction op into the binary. That's the whole challenge here.  We can't expose templare functions through the public APIs in cuVS.

Reply:
Ok what I could do is have an enum - such as "LINKAGE_TYPE" or "DISTANCE_TYPE", where one option can be L2 and the other can be mutual reachability. If it is mutual reachability I can keep the reduction op inside cuvs and call that.

Reply:
Moved everything to the reachability header in the helpers namespace. Removed the template reduction op.